docs/4.4/api-reference.md

```diff
@@ -11,57 +11,326 @@ Teleport is currently working on documenting our API.
+
+```yaml
+# Copy and Paste the below and run on the Teleport Auth server.
+$ cat > api-admin.yaml <<EOF
+{!examples/go-client/api-admin.yaml!}
 EOF
-# ...
-$ tctl create rscs.yaml
-# ...
-$ tctl auth sign --format=tls --user=access-plugin --out=auth
-# ...
+
+$ tctl create -f api-admin.yaml
+$ mkdir -p certs
+$ tctl auth sign --format=tls --user=api-admin --out=certs/api-admin
 ```
 
-The above sequence should result in three PEM encoded files being generated:
-`auth.crt`, `auth.key`, and `auth.cas` (certificate, private key, and CA certs respectively).
+This should result in three PEM encoded files being generated in the `/certs` directory: `api-admin.crt`, `api-admin.key`, and `api-admin.cas` (certificate, private key, and CA certs respectively).
 
-Note: by default, tctl auth sign produces certificates with a relatively short lifetime.
-For production deployments, the --ttl flag can be used to ensure a more practical
-certificate lifetime.
+Move the `/certs` folder into your `/api-examples` folder.
 
-# gRPC APIs
+!!! Note
+    By default, `tctl auth sign` produces certificates with a relatively short lifetime.
+    For production deployments, the `--ttl` flag can be used to ensure a more practical
+    certificate lifetime.
+
+## Go Client
+
+Add `client.go` into `/api-examples`.
+
+**client.go**
+
+```go
+{!examples/go-client/client.go!}
+```
+
+## Go Packages
+
+Copy the Teleport module's go.mod below into `/api-examples` and then run `go mod tidy` to slim it down to only what's needed for these api examples.
+
+```
+{!go.mod!}
+```
+
+## Main file
+
+Add this main file to your `/api-examples` folder. Now you can simply plug in the examples below and then run `go run .` to see them in action.
+
+**main.go**
+
+```go
+package main
+
+import (
+	"fmt"
+	"log"
+)
+
+func main() {
+	log.Printf("Starting teleport client...")
+	client, err := connectClient()
+	if err != nil {
+		log.Fatalf("Failed to create client: %v", err)
+	}
+}
+```
+
+## Roles
+
+[Roles](enterprise/ssh-rbac.md#roles) are used to define what resources and actions a user is allowed/denied access to. It's best to make your role's permissions as stringent as possible.
+
+**Retrieve role**
+
+```go
+role, err := client.GetRole("auditor")
+if err != nil {
+  return err
+}
+```
+
+**Create a new role**
+
+```go
+// create a new auditor role which has very limited permissions
+role, err := services.NewRole("auditor", services.RoleSpecV3{
+  Options: services.RoleOptions{
+    MaxSessionTTL: services.Duration(time.Hour),
+  },
+  Allow: services.RoleConditions{
+    Logins: []string{"auditor"},
+    Rules: []services.Rule{
+      // - resources: ['session']
+      //   verbs: ['list', 'read']
+      services.NewRule(services.KindSession, services.RO()),
+    },
+  },
+  Deny: services.RoleConditions{
+    // node_labels: '*': '*'
+    NodeLabels: services.Labels{"*": []string{"*"}},
+  },
+})
+if err != nil {
+  return err
+}
+
+if err = client.UpsertRole(ctx, role); err != nil {
```

Comment:
Let's use Create and Update whenever possible in our example. Upsert was a mistake and we should be deprecating it